# Building a database for Crime Reports

In this project, we will build a database for storing data related with crimes that occurred in Boston. We'll be using Postgres.

### Goals

The goal of this guided project is for you to create a database named `crimes_db` with a table – `boston_crimes` – with appropriate datatypes for storing the data from the `boston.csv` file. 

We will be creating the table inside a schema named `crimes`. 

We will also create the `readonly` and `readwrite` groups with the appropriate privileges as we have discussed in this course. Finally, you will also need to create one user for each of these groups.

<img src="goal.png">

### Creating Crimes Database

Let's start by creating the `crimes_db` database. We'll first connect to the `dq` database with user `dq` when connecting to the postgres server. We'll create our database, then connect to and setup the scheme - `crimes`

In [1]:
import psycopg2

In [71]:
# Connect to postgres database
conn = psycopg2.connect(dbname = "postgres", user = "postgres")
conn.autocommit = True
cur = conn.cursor()

In [ ]:
# Create crimes database
cur.execute("CREATE DATABASE crimes_db;")
conn.autocommit = False

In [72]:
conn.close()

In [2]:
# Connect to our newley created database
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")
cur = conn.cursor()

In [81]:
# Create a schema within our database

cur.execute("CREATE SCHEMA crimes;")

### Obtaining Column names and sample

Let's take a look at our Boston data

In [3]:
import csv

In [4]:
# Print header and first two rows of our data
with open("Boston.csv", "r") as file:
    reader = csv.reader(file)
    col_headers = next(reader)
    first_row = next(reader)
    print(col_headers)
    print(first_row)

['incident_number', 'offense_code', 'description', 'date', 'day_of_the_week', 'lat', 'long']
['1', '619', 'LARCENY ALL OTHERS', '2018-09-02', 'Sunday', '42.35779134', '-71.13937053']


### Creating an Auxillary Function

Let's create a function that will tell us all the different values for each column in our data

In [5]:
def get_col_set(csv_filename, col_index):
    values = set()
    with open(csv_filename, 'r') as file:
        next(file)
        reader = csv.reader(file)
        for row in reader:
            values.add(row[col_index])
    return values

In [6]:
print(len(col_headers))

7


In [7]:
for index in range(len(col_headers)):
    values = get_col_set("Boston.csv", index)
    print(f'{col_headers[index]:20} {len(values)}')

incident_number      298329
offense_code         219
description          239
date                 1177
day_of_the_week      7
lat                  18177
long                 18177


### Finding the maximum length

Let's compute the maximum length of each value in the `description` column and `offense_code` column.

In [8]:
description_values = get_col_set("Boston.csv", 2)

max_len = 0
for description in description_values:
    if len(description) > max_len:
        max_len = len(description)
print(max_len)

58


In [9]:
offense_values = get_col_set("Boston.csv", 1)

max_len = 0
for code in offense_values:
    if len(code) > max_len:
        max_len = len(code)
print(max_len)

4


### Creating the Table

Let's now create a table to store the Boston crime data. We can see that for `day_of_the_week` there are 7 unique values. For this, we might want to use an `enumerate datatype`. Let's begin.

In [10]:
print(get_col_set("Boston.csv", 4))

{'Friday', 'Monday', 'Saturday', 'Tuesday', 'Sunday', 'Wednesday', 'Thursday'}


In [82]:
cur.execute("""CREATE TYPE days AS ENUM 
                ('Saturday', 'Wednesday', 'Thursday', 'Sunday', 'Friday', 'Monday', 'Tuesday');""")

In [11]:
print(col_headers)

['incident_number', 'offense_code', 'description', 'date', 'day_of_the_week', 'lat', 'long']


In [12]:
print(first_row)

['1', '619', 'LARCENY ALL OTHERS', '2018-09-02', 'Sunday', '42.35779134', '-71.13937053']


For these remaining columns, what datatypes should we use?

* `incident_number` - Because numbers go up to 298329, we'll use the `integer` datatype


* `offense_code` - We'll also go with the `smallint` datatype here


* `description` - As this is text, and the maximum length is 58. We'll round up and use the `varchar` datatype with a value of 60.


* `date` - We'll use the `date` type here, which will allow us to perform sane queries.


* `day_of_the_week` - We'll go with our `enumerated data` type. Why? Because because we have a predefined set of values. When we create an enumerated datatype, each possible value is assigned to a 4 byte index within a seperate table. This will ultimately save us memory.


* `lat` - Because we're dealing with decimals, we'll go with the `decimal` datatype


* `long` - Same as above

In [83]:
# Create our boston_crimes table within our new schema
try:
    cur.execute("""CREATE TABLE crimes.boston_crimes (
                incident_number integer PRIMARY KEY,
                 offense_code smallint,
                 description varchar(60),
                 date date,
                 day_of_the_week days,
                 lat decimal,
                 long decimal);""")
except Exception as ex:
    print(ex)

In [13]:
cur.execute("SELECT * FROM crimes.boston_crimes LIMIT 10;")

In [14]:
for i in cur.description:
    print(i)

Column(name='incident_number', type_code=23)
Column(name='offense_code', type_code=21)
Column(name='description', type_code=1043)
Column(name='date', type_code=1082)
Column(name='day_of_the_week', type_code=16468)
Column(name='lat', type_code=1700)
Column(name='long', type_code=1700)


In [15]:
conn.commit()
conn.close()

### Loading the Data

Let's now copy our Boston data into the database. We'll use the `copy_expert` method as below.

In [16]:
# Connect again to our database
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")
cur = conn.cursor()

In [21]:
with open("Boston.csv", "r") as f:
    cur.copy_expert("COPY crimes.boston_crimes FROM STDIN WITH CSV HEADER;", f)

In [17]:
cur.execute("SELECT * FROM pg_user;")

In [18]:
cur.fetchall()

[('postgres', 10, True, True, True, True, '********', None, None),
 ('aaronwright', 16384, True, True, False, False, '********', None, None)]

### Revoking Public Privileges

When we created our database `crimes_db` a `public` schema was automatically created, and grants access to this schema via a role/group `public`. 

We want to revoke all privliges from the `public` group on the `public` schema.

We also want to revoke all privliges of `public` on the `crimes_db` database.

If we don't do this, a user will inherit privileges from the `public` role even if we restrict their privileges and will be able to do stuff with the `public` schema (e.g. create objects)

In [99]:
cur.execute("REVOKE ALL ON SCHEMA public FROM public;")

In [100]:
cur.execute("REVOKE ALL ON DATABASE crimes_db FROM public;")

### Creating User Groups

Let's now setup some groups. Mainly we want `readonly` and `readwrite` groups setup.

In [26]:
# Create our groups/roles
cur.execute("CREATE GROUP readonly NOLOGIN;")
cur.execute("CREATE GROUP readwrite NOLOGIN;")

In [27]:
# Grant our groups/roles connections prviliges to our database
cur.execute("GRANT CONNECT ON DATABASE crimes_db TO readonly;")
cur.execute("GRANT CONNECT ON DATABASE crimes_db TO readwrite;")

In [28]:
# Grant usage privilages to our schema
cur.execute("GRANT USAGE ON SCHEMA crimes TO readonly;")
cur.execute("GRANT USAGE ON SCHEMA crimes TO readwrite")

In [31]:
# Grant prvilieges to our groups
cur.execute("GRANT SELECT ON ALL TABLES IN SCHEMA CRIMES TO readonly;")
cur.execute("GRANT SELECT, INSERT, DELETE, UPDATE ON ALL TABLES IN SCHEMA CRIMES TO readwrite;")

In [34]:
# Check our privileges are correct
cur.execute("SELECT grantor, grantee, privilege_type FROM information_schema.table_privileges WHERE table_name = 'boston_crimes';")

In [35]:
cur.fetchall()

[('postgres', 'postgres', 'INSERT'),
 ('postgres', 'postgres', 'SELECT'),
 ('postgres', 'postgres', 'UPDATE'),
 ('postgres', 'postgres', 'DELETE'),
 ('postgres', 'postgres', 'TRUNCATE'),
 ('postgres', 'postgres', 'REFERENCES'),
 ('postgres', 'postgres', 'TRIGGER'),
 ('postgres', 'readonly', 'SELECT'),
 ('postgres', 'readwrite', 'INSERT'),
 ('postgres', 'readwrite', 'SELECT'),
 ('postgres', 'readwrite', 'UPDATE'),
 ('postgres', 'readwrite', 'DELETE')]

### Creating Users

Let's now create one user in each group, giving them a password.

In [36]:
# Set up users
cur.execute("CREATE USER data_analyst WITH PASSWORD 'secret1';")
cur.execute("CREATE USER data_scientist WITH PASSWORD 'secret2';")

In [37]:
# Assign users to groups
cur.execute("GRANT readonly TO data_analyst;")
cur.execute("GRANT readwrite TO data_scientist;")

In [38]:
conn.commit()
conn.close()

### Testing

Let's now test we've set everything up correctly

In [42]:
conn = psycopg2.connect(dbname = "crimes_db", user = "postgres")

In [43]:
cur = conn.cursor()

In [76]:
# Check our roles have the right prvileges
cur.execute("""SELECT grantee, privilege_type
                FROM information_schema.table_privileges
                WHERE grantee = 'readwrite' OR grantee = 'readonly';""")
cur.fetchall()

[('readonly', 'SELECT'),
 ('readwrite', 'INSERT'),
 ('readwrite', 'SELECT'),
 ('readwrite', 'UPDATE'),
 ('readwrite', 'DELETE')]

In [86]:
conn.rollback()